yarpl/include/yarpl/observable/Observable.h

```diff
@@ -167,6 +203,57 @@ Reference<Observable<T>> Observable<T>::subscribeOn(folly::Executor& executor) {
 }
 
 template <typename T>
+template <typename Function>
+Reference<Observable<T>> Observable<T>::doOnSubscribe(Function function) {
+  return details::createDoOperator(ref_from_this(this), std::move(function), [](const T&){}, [](auto){}, []{});
+}
+
+template <typename T>
+template <typename Function>
+Reference<Observable<T>> Observable<T>::doOnNext(Function function) {
+  return details::createDoOperator(ref_from_this(this), []{}, std::move(function), [](auto){}, []{});
+}
+
+template <typename T>
+template <typename Function>
+Reference<Observable<T>> Observable<T>::doOnError(Function function) {
+  return details::createDoOperator(ref_from_this(this), []{}, [](const T&){}, std::move(function), []{});
+}
+
+template <typename T>
+template <typename Function>
+Reference<Observable<T>> Observable<T>::doOnComplete(Function function) {
+  return details::createDoOperator(ref_from_this(this), []{}, [](const T&){}, [](auto){}, std::move(function));
+}
+
+template <typename T>
+template <typename Function>
+Reference<Observable<T>> Observable<T>::doOnTerminate(Function function) {
```

Comment:
Terminate won't be called in case the observable/flowable is `cancel()`'d, which is different than what SafeSubscriber does. We should consider the meaning of `terminate` in yarpl, and if that should exclude the case where `cancel()` is called so we're consistent.

Reply:
I will add a comment there to review it after we have SafeSubscriber/SubscriberBase